status-im/status-react
# [#13964 #13962] Display contact requests

fixes #13962
fixes #13964

### Summary

This PR adds support for Contact Request notifications in the new Activity Center (read-only at the moment, i.e. non-interactive).

<table>
  <thead>
    <tr>
      <th>
        <img alt="Screenshot" src="https://user-images.githubusercontent.com/46027/190483706-ec2f9ccf-c9e3-497e-9f8a-8a2ad66f2eee.png" height="600"/>
        <p><strong>Contact requests in various states</strong></p>
      </th>
    </tr>
  </thead>
</table>

### Review notes

- Upgrades clj-kondo to the latest version. The original version was pretty old, and it was the cause of incorrect static analysis that generated errors in the CI. For an example, see these [logs in Jenkins]( https://ci.status.im/blue/rest/organizations/jenkins/pipelines/status-mobile/pipelines/prs/pipelines/tests/branches/PR-14014/runs/4/nodes/35/log/?start=0).
- The user can't yet interact with notifications. See some of the remaining issues in https://github.com/status-im/status-mobile/labels/Activity%20center.
- The new `status-tag` component was improved to accept custom labels, otherwise it was only capable of displaying fixed text, like "positive", "negative", etc.
- Figma flows https://www.figma.com/file/eDfxTa9IoaCMUy5cLTp0ys/Shell-for-Mobile?node-id=639%3A21926
- The `utils.datetime` namespace was reformatted to use the widely accepted convention in Lisps (and Clojure) to separate top-level sections with 3 or 4 backticks. See (https://github.com/bbatsov/clojure-style-guide#heading-comments). On a side note, if you use Emacs, you can enable the `outline-minor-mode` to automatically get collapsible sections (as in Org mode).

<table>
  <thead>
    <tr>
      <th>
        <img alt="Screenshot" src="https://user-images.githubusercontent.com/46027/190483131-c3f45ba5-e4f4-41bf-a842-6c6674a136f7.jpg" width="500"/>
        <p><strong>Lisp heading comments can be collapsed in Emacs</strong></p>
      </th>
    </tr>
  </thead>
</table>

### Testing notes

Just like the merged PR https://github.com/status-im/status-mobile/pull/13987, this one doesn't yet replace the old Notifications Center with the new Activity Center. This will only be done once the Status Go [draft branch](https://github.com/status-im/status-go/tree/feature/activity-center-read-unread) is merged, but there are still some outstanding tests to be done.

Meanwhile, you need to do the following:

- Checkout status-go branch `feature/activity-center-read-unread`.
- Always run `export STATUS_GO_SRC_OVERRIDE=$HOME/<path-to-your-status-go-repo>` before running targets such as `make run-android`.
- Apply the following patch to status-mobile:

```diff
modified   src/status_im/ui/screens/home/views.cljs
@@ -270,7 +270,7 @@
                           :accessibility-label :notifications-button
                           :on-press #(do
                                        (re-frame/dispatch [:mark-all-activity-center-notifications-as-read])
-                                       (re-frame/dispatch [:navigate-to :notifications-center]))}
+                                       (re-frame/dispatch [:navigate-to :activity-center]))}
       [icons/icon :main-icons/notification2 {:color (quo2.colors/theme-colors quo2.colors/black quo2.colors/white)}]]
      (when (pos? notif-count)
        [react/view {:style (merge (styles/counter-public-container) {:top 5 :right 5})
```

### Steps to test

You can follow the use cases below if you'd like to do some manual testing. Please, don't consider this an exaustive list of use cases. In all cases, Bob has enabled _accept chat from anyone_ in his app settings.

>Given that Alice has not added Bob as a contact,
>When Alice opens a chat with Bob by using his chat key,
>And Alice sends her first message to him,
>Then Bob sees the last message from Alice in his Activity Center (aka 1:1 notification).

>Given that Alice already sent a message to Bob,
>When Alice opens the chat with Bob,
>And when Alice clicks on Add Contact,
>Then Bob receives a notification in the Activity Center,
>And Bob can either accept or decline the contact request.

>Given that Bob accepted a contact request from Alice,
>When Bob removes Alice from his contact list,
>Then Bob doesn't see notifications related to Alice in the Activity Center.

>Given that Bob accepted a contact request from Alice,
>When Bob opens the Activity Center,
>Then he sees a notification saying he accepted the contact request from Alice.

>Given that Bob declined a contact request from Alice,
>When Bob opens the Activity Center,
>Then he sees a notification saying he declined the contact request from Alice.

>Given that Bob gave a nickname to Alice (only he sees),
>When Bob opens the Activity Center,
>Then he sees all references to Alice replaced with her nickname.

status: ready
